File: fabric-sdk-go/crypto/crypto.go
// Package crypto provides the cryptographic primitives used by the Private Mesh
// fabric SDK: XChaCha20-Poly1305 AEAD for payload and FIF body encryption, plus
// the KDFs (Argon2id, HKDF-SHA256) defined in fabric-spec-001-v1.0.md.
//
// The primitives here are thin wrappers over golang.org/x/crypto. They exist to
// give the rest of the SDK a stable, fabric-shaped API and to centralize the
// "always XChaCha (NewX), always 32-byte keys, always 24-byte nonces" choices
// so they cannot drift.
package crypto

import (
	"crypto/rand"
	"errors"
	"fmt"

	"golang.org/x/crypto/chacha20poly1305"
)

// KeySize is the symmetric key length used everywhere in the fabric (32 bytes).
const KeySize = chacha20poly1305.KeySize

// NonceSize is the XChaCha20-Poly1305 nonce length (24 bytes).
const NonceSize = chacha20poly1305.NonceSizeX

// TagSize is the Poly1305 authentication tag length (16 bytes).
const TagSize = 16

// SaltSize is the FIF envelope salt length (16 bytes per fabric-spec).
const SaltSize = 16

// ErrInvalidKeySize is returned by Seal/Open when the key is not KeySize bytes.
var ErrInvalidKeySize = errors.New("fabric/crypto: key must be 32 bytes")

// ErrInvalidNonceSize is returned by Seal/Open when the nonce is not NonceSize bytes.
var ErrInvalidNonceSize = errors.New("fabric/crypto: nonce must be 24 bytes")

// Seal encrypts plaintext with XChaCha20-Poly1305. The returned slice contains
// the ciphertext followed by the 16-byte Poly1305 tag. aad is authenticated
// but not encrypted; pass nil if there is no AAD.
func Seal(key, nonce, plaintext, aad []byte) ([]byte, error) {
	if len(key) != KeySize {
		return nil, ErrInvalidKeySize
	}
	if len(nonce) != NonceSize {
		return nil, ErrInvalidNonceSize
	}
	aead, err := chacha20poly1305.NewX(key)
	if err != nil {
		return nil, fmt.Errorf("fabric/crypto: aead init: %w", err)
	}
	return aead.Seal(nil, nonce, plaintext, aad), nil
}

// Open decrypts and authenticates ciphertext produced by Seal. The ciphertext
// slice MUST include the 16-byte Poly1305 tag appended by Seal. Returns the
// plaintext or an error if authentication fails.
func Open(key, nonce, ciphertext, aad []byte) ([]byte, error) {
	if len(key) != KeySize {
		return nil, ErrInvalidKeySize
	}
	if len(nonce) != NonceSize {
		return nil, ErrInvalidNonceSize
	}
	aead, err := chacha20poly1305.NewX(key)
	if err != nil {
		return nil, fmt.Errorf("fabric/crypto: aead init: %w", err)
	}
	return aead.Open(nil, nonce, ciphertext, aad)
}

// RandomBytes returns n cryptographically-random bytes.
func RandomBytes(n int) ([]byte, error) {
	buf := make([]byte, n)
	if _, err := rand.Read(buf); err != nil {
		return nil, fmt.Errorf("fabric/crypto: random: %w", err)
	}
	return buf, nil
}

// RandomNonce returns 24 fresh random bytes suitable for an XChaCha20-Poly1305 nonce.
func RandomNonce() ([]byte, error) { return RandomBytes(NonceSize) }

// RandomSalt returns 16 fresh random bytes suitable for an Argon2id salt.
func RandomSalt() ([]byte, error) { return RandomBytes(SaltSize) }

File: fabric-sdk-js/src/crypto.js
// Cryptographic primitives used by the Private Mesh fabric SDK.
//
// Authoritative wire details: docs/specs/fabric-spec-001-v1.0.md §"Encryption"
// and §"Fabric Identity File". Defaults here match the Go SDK
// (fabric-sdk-go/crypto) so artifacts produced by either side round-trip.

import { xchacha20poly1305 } from '@noble/ciphers/chacha';
import { argon2id } from 'hash-wasm';

/** Symmetric key length used everywhere (32 bytes). */
export const KEY_SIZE = 32;
/** XChaCha20-Poly1305 nonce length (24 bytes). */
export const NONCE_SIZE = 24;
/** Poly1305 authentication tag length (16 bytes). */
export const TAG_SIZE = 16;
/** FIF envelope salt length (16 bytes per fabric-spec). */
export const SALT_SIZE = 16;

/** Default Argon2id params for FIF passphrase-only envelopes (matches Go SDK). */
export const DEFAULT_ARGON2ID_PARAMS = Object.freeze({
  time: 3, // t_cost
  memory: 65536, // m_cost in KiB
  parallelism: 4,
  hashLength: KEY_SIZE,
});

/**
 * Encrypt plaintext under XChaCha20-Poly1305.
 *
 * @param {Uint8Array} key   32-byte key.
 * @param {Uint8Array} nonce 24-byte nonce.
 * @param {Uint8Array} plaintext
 * @param {Uint8Array|null} aad Additional authenticated data (or null).
 * @returns {Uint8Array} ciphertext concatenated with the 16-byte tag.
 */
export function seal(key, nonce, plaintext, aad) {
  if (key.length !== KEY_SIZE) throw new Error('fabric/crypto: key must be 32 bytes');
  if (nonce.length !== NONCE_SIZE) throw new Error('fabric/crypto: nonce must be 24 bytes');
  const cipher = xchacha20poly1305(key, nonce, aad ?? undefined);
  return cipher.encrypt(plaintext);
}

/**
 * Decrypt and authenticate a Seal output. Throws on authentication failure.
 *
 * @param {Uint8Array} key
 * @param {Uint8Array} nonce
 * @param {Uint8Array} ciphertext  ciphertext || tag, as produced by seal.
 * @param {Uint8Array|null} aad
 * @returns {Uint8Array} plaintext
 */
export function open(key, nonce, ciphertext, aad) {
  if (key.length !== KEY_SIZE) throw new Error('fabric/crypto: key must be 32 bytes');
  if (nonce.length !== NONCE_SIZE) throw new Error('fabric/crypto: nonce must be 24 bytes');
  const cipher = xchacha20poly1305(key, nonce, aad ?? undefined);
  return cipher.decrypt(ciphertext);
}

/** Return n cryptographically-random bytes using WebCrypto. */
export function randomBytes(n) {
  const out = new Uint8Array(n);
  globalThis.crypto.getRandomValues(out);
  return out;
}

/** 24 fresh random bytes suitable for XChaCha20-Poly1305 nonces. */
export const randomNonce = () => randomBytes(NONCE_SIZE);
/** 16 fresh random bytes suitable for Argon2id salts. */
export const randomSalt = () => randomBytes(SALT_SIZE);

/**
 * Derive a 32-byte symmetric key from a passphrase using Argon2id.
 *
 * @param {string} passphrase
 * @param {Uint8Array} salt
 * @param {{time: number, memory: number, parallelism: number, hashLength: number}} [params]
 * @returns {Promise<Uint8Array>} 32-byte derived key.
 */
export async function deriveKeyArgon2id(passphrase, salt, params = DEFAULT_ARGON2ID_PARAMS) {
  const result = await argon2id({
    password: passphrase,
    salt,
    iterations: params.time,
    memorySize: params.memory,
    parallelism: params.parallelism,
    hashLength: params.hashLength,
    outputType: 'binary',
  });
  return result instanceof Uint8Array ? result : new Uint8Array(result);
}

/**
 * Derive a 32-byte symmetric key via HKDF-SHA256.
 *
 * @param {Uint8Array} secret
 * @param {Uint8Array} salt   May be empty; HKDF treats empty as zero-string of HashLen.
 * @param {Uint8Array} info
 * @returns {Promise<Uint8Array>}
 */
export async function deriveKey(secret, salt, info) {
  const key = await globalThis.crypto.subtle.importKey(
    'raw',
    secret,
    { name: 'HKDF' },
    false,
    ['deriveBits'],
  );
  const bits = await globalThis.crypto.subtle.deriveBits(
    { name: 'HKDF', hash: 'SHA-256', salt, info },
    key,
    KEY_SIZE * 8,
  );
  return new Uint8Array(bits);
}
